main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1071 @@
+        val refType = Type.mkRef(elmVar, regionVar, loc)
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        c.expectTypeM(expected = refType, actual = tpe1, exp1.loc)
+        c.expectTypeM(expected = elmVar, actual = tpe2, exp2.loc)
+        c.unifyTypeM(evar, Type.mkUnion(eff1, eff2, regionVar, loc), loc)
+        val resTpe = Type.Unit
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.Ascribe(exp, expectedTpe, expectedEff, tvar, loc) =>
+        // An ascribe expression is sound; the type system checks that the declared type matches the inferred type.
+        val (actualTpe, actualEff) = visitExp(exp)
+        c.expectTypeM(expected = expectedTpe.getOrElse(Type.freshVar(Kind.Star, loc)), actual = actualTpe, loc)
+        c.unifyTypeM(actualTpe, tvar, loc)
+        c.expectTypeM(expected = expectedEff.getOrElse(Type.freshVar(Kind.Eff, loc)), actual = actualEff, loc)
+        val resTpe = tvar
+        val resEff = actualEff
+        (resTpe, resEff)
+
+      case Expr.InstanceOf(exp, clazz, loc) =>
+        val (tpe, eff) = visitExp(exp)
+        c.expectTypeM(expected = Type.Pure, actual = eff, exp.loc)
+        val resTpe = Type.Bool
+        val resEff = Type.Pure
+        (resTpe, resEff)
+
+      case Expr.CheckedCast(cast, exp, tvar, evar, loc) =>
+        cast match {
+          case Ast.CheckedCastType.TypeCast =>
+            // Ignore the inferred type of exp.
+            val (_, eff) = visitExp(exp)
+            c.unifyTypeM(evar, eff, loc)
+            val resTpe = tvar
+            val resEff = evar
+            (resTpe, resEff)
+
+          case Ast.CheckedCastType.EffectCast =>
+            // We simply union the purity and effect with a fresh variable.
+            val (tpe, eff) = visitExp(exp)
+            c.unifyTypeM(tvar, tpe, loc)
+            val resTpe = tvar
+            val resEff = Type.mkUnion(eff, evar, loc)
+            (resTpe, resEff)
+        }
+
+      case Expr.UncheckedCast(exp, declaredTpe, declaredEff, tvar, loc) =>
+        // A cast expression is unsound; the type system assumes the declared type is correct.
+        val (actualTyp, actualEff) = visitExp(exp)
+        c.unifyTypeM(tvar, declaredTpe.getOrElse(actualTyp), loc)
+        val resTpe = tvar
+        val resEff = declaredEff.getOrElse(actualEff)
+        (resTpe, resEff)
+
+      case Expr.UncheckedMaskingCast(exp, loc) =>
+        val (tpe, eff) = visitExp(exp)
+        val resTpe = tpe
+        val resEff = Type.Pure
+        (resTpe, resEff)
+
+      case Expr.Without(exp, effUse, loc) =>
+        val effType = Type.Cst(TypeConstructor.Effect(effUse.sym), effUse.loc)
+        //        val expected = Type.mkDifference(Type.freshVar(Kind.Bool, loc), effType, loc)
+        // TODO EFF-MIGRATION use expected
+        val (tpe, eff) = visitExp(exp)
+        val resTpe = tpe
+        val resEff = eff
+        (resTpe, resEff)
+
+      case Expr.TryCatch(exp, rules, loc) =>
```

Comment:
Lets add a visitTryCatchRule

Reply:
could also add `visitRules` a la my suggestion in match so the main snippet here is more succinct